File: leetcode/medium/Q3_LongestSubstringWithoutRepeatingCharacters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        max_num = 0
        sets = []
        for c in s:
            if c in sets:
                while c in sets:
                    sets.pop(0)
            sets.append(c)
            max_num = max(max_num, len(sets))
        return max_num

File: src/algorithm1/slidingwindow/Q3_LongestSubstringWithoutRepeatingCharacters.java
package algorithm1.slidingwindow;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;

public class Q3_LongestSubstringWithoutRepeatingCharacters {
    class Solution {
        public int lengthOfLongestSubstring(String s) {
            char[] charArr = s.toCharArray();
            int start = 0;
            int end = 0;
            Set<Character> characterSet = new HashSet<>();
            int maxLength = 0;
            for (int i = 0; i < charArr.length; i++) {
                while (characterSet.contains(charArr[i])){
                    characterSet.remove(charArr[start++]);
                }
                characterSet.add(charArr[i]);
                end += 1;
                maxLength = Math.max(maxLength, end - start);
            }
            return maxLength;
        }
    }
}
